rust/relay/ebpf-turn-router/src/main.rs

```diff
@@ -56,40 +56,53 @@ static UDP_TO_CHAN_66: HashMap<PortAndPeerV6, ClientAndChannelV6> =
 
 #[xdp]
 pub fn handle_turn(ctx: XdpContext) -> u32 {
-    try_handle_turn(&ctx).unwrap_or_else(|e| {
-        let action = e.xdp_action();
-
-        debug!(&ctx, "Failed to handle packet {}; action = {}", e, action);
-
-        action
+    try_handle_turn(&ctx).unwrap_or_else(|e| match e {
+        Error::NotUdp
+        | Error::NotTurn
+        | Error::NotIp
+        | Error::NotAChannelDataMessage
+        | Error::Ipv4PacketWithOptions => xdp_action::XDP_PASS,
+
+        Error::XdpStoreBytesFailed
+        | Error::XdpAdjustHeadFailed
+        | Error::XdpLoadBytesFailed
+        | Error::PacketTooShort
+        | Error::NoChannelBinding => {
+            debug!(&ctx, "Failed to handle packet: {}", e);
+
+            xdp_action::XDP_PASS
+        }
+
+        Error::BadChannelDataLength => {
+            debug!(&ctx, "Failed to handle packet: {}; dropping", e);
+
+            xdp_action::XDP_DROP
+        }
     })
 }
 
 #[inline(always)]
 fn try_handle_turn(ctx: &XdpContext) -> Result<u32, Error> {
     let eth = Eth::parse(ctx)?;
 
-    let action = match eth.ether_type() {
+    match eth.ether_type() {
         EtherType::Ipv4 => try_handle_turn_ipv4(ctx)?,
         EtherType::Ipv6 => try_handle_turn_ipv6(ctx)?,
-        _ => return Ok(xdp_action::XDP_PASS),
+        _ => return Err(Error::NotIp),
     };
 
-    // If we send the packet back out, swap the source and destination MAC addresses.
-    // We will have adjusted the packet pointers so we need to reparse the packet.
-    if action == xdp_action::XDP_TX {
-        Eth::parse(ctx)?.swap_src_and_dst();
-    }
+    // If we get to here, we modified the packet and need to send it back out again.
+    Eth::parse(ctx)?.swap_src_and_dst();
 
-    Ok(action)
+    Ok(xdp_action::XDP_TX)
```

Comment:
This is the main change! Instead of passing back an action through mulitple layers, we simply assume that if we make it to this point, we've successfully modified the packet and want to send it back out.

To achieve this, all other codepaths are now using errors.